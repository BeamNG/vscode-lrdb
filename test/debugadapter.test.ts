//
// Note: This example test is leveraging the Mocha test framework.
// Please refer to their documentation on https://mochajs.org/ for help.
//

// The module 'assert' provides assertion methods from node
import * as assert from 'assert';
import * as path from 'path';

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import * as vscode from 'vscode';
import { DebugClient } from 'vscode-debugadapter-testsupport';
import { DebugProtocol } from 'vscode-debugprotocol';

// Defines a Mocha test suite to group tests of similar kind together
suite("Lua Debug Adapter", () => {
	const DEBUG_ADAPTER = './out/src/lrdbDebug.js';
	const PROJECT_ROOT = path.join(__dirname, '../../');
	const DATA_ROOT = path.join(PROJECT_ROOT, 'test/lua/');


	let dc: DebugClient;

	setup(() => {
		dc = new DebugClient('node', DEBUG_ADAPTER, 'lua');

		return dc.start();
	});

	teardown(() => dc.stop());


	suite('basic', () => {
		test('unknown request should produce error', done => {
			dc.send('illegal_request').then(() => {
				done(new Error("does not report error on unknown request"));
			}).catch(() => {
				done();
			});
		});
	});


	suite('initialize', () => {

		test('should return supported features', () => {
			return dc.initializeRequest().then(response => {
				assert.equal(response.body.supportsConfigurationDoneRequest, true);
			});
		});

		test('should produce error for invalid \'pathFormat\'', done => {
			dc.initializeRequest({
				adapterID: 'lua',
				linesStartAt1: true,
				columnsStartAt1: true,
				pathFormat: 'url'
			}).then(response => {
				done(new Error("does not report error on invalid 'pathFormat' attribute"));
			}).catch(err => {
				// error expected
				done();
			});
		});
	});


	suite('launch', () => {

		test('should run program to the end', () => {
			const PROGRAM = path.join(DATA_ROOT, 'loop_test.lua');

			return Promise.all([
				dc.launch({ program: PROGRAM }),
				dc.configurationSequence(),
				dc.waitForEvent('terminated')
			]);
		});

		test('should stop on entry', () => {

			const PROGRAM = path.join(DATA_ROOT, 'loop_test.lua');
			const ENTRY_LINE = 1;
			return Promise.all([
				dc.launch({ program: PROGRAM, stopOnEntry: true }),
				dc.configurationSequence(),
				dc.waitForEvent('stopped'),
				dc.assertStoppedLocation('entry', { line: ENTRY_LINE } )
			]);
		});
	});
	
	suite('breakpoint', () => {
		test('should stop on breakpoint', () => {
			const PROGRAM = path.join(DATA_ROOT, 'loop_test.lua');
			const BREAK_LINE = 5;
			return Promise.all([
				dc.hitBreakpoint({ program: PROGRAM},{path:PROGRAM,line:BREAK_LINE}),
				dc.configurationSequence(),
				dc.waitForEvent('stopped'),
				dc.assertStoppedLocation('breakpoint', { line: BREAK_LINE } ),
			]);
		});
	});
});