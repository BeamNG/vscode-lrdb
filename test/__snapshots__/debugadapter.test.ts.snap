// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lua Debug Adapter global check global values 1`] = `
Object {
  "body": Object {
    "variables": Array [
      Object {
        "name": "_G",
        "type": "object",
        "value": "{\\"_G\\":{\\"table\\":\\"0x0\\"},\\"_VERSION\\":\\"Lua 5.3\\",\\"assert\\":\\"function: 0x67\\",\\"collectgarbage\\":\\"function: 0x68\\",\\"coroutine\\":{\\"table\\":\\"0x0\\"},\\"debug\\":{\\"table\\":\\"0x0\\"},\\"dofile\\":\\"function: 0x69\\",\\"error\\":\\"function: 0x6a\\",\\"getmetatable\\":\\"function: 0x6b\\",\\"global_value\\":1,\\"io\\":{\\"table\\":\\"0x0\\"},\\"ipairs\\":\\"function: 0x6c\\",\\"load\\":\\"function: 0x6e\\",\\"loadfile\\":\\"function: 0x6d\\",\\"math\\":{\\"table\\":\\"0x0\\"},\\"next\\":\\"function: 0x65\\",\\"os\\":{\\"table\\":\\"0x0\\"},\\"package\\":{\\"table\\":\\"0x0\\"},\\"pairs\\":\\"function: 0x6f\\",\\"pcall\\":\\"function: 0x70\\",\\"print\\":\\"function: 0x71\\",\\"rawequal\\":\\"function: 0x72\\",\\"rawget\\":\\"function: 0x74\\",\\"rawlen\\":\\"function: 0x73\\",\\"rawset\\":\\"function: 0x75\\",\\"require\\":\\"function: 0x50bfe0\\",\\"select\\":\\"function: 0x76\\",\\"setmetatable\\":\\"function: 0x77\\",\\"string\\":{\\"table\\":\\"0x0\\"},\\"table\\":{\\"table\\":\\"0x0\\"},\\"tonumber\\":\\"function: 0x78\\",\\"tostring\\":\\"function: 0x79\\",\\"type\\":\\"function: 0x7a\\",\\"utf8\\":{\\"table\\":\\"0x0\\"},\\"xpcall\\":\\"function: 0x7b\\"}",
        "variablesReference": 1003,
      },
      Object {
        "name": "_VERSION",
        "type": "string",
        "value": "\\"Lua 5.3\\"",
        "variablesReference": 0,
      },
      Object {
        "name": "assert",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "collectgarbage",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "coroutine",
        "type": "object",
        "value": "{\\"create\\":\\"function: 0x7d\\",\\"isyieldable\\":\\"function: 0x83\\",\\"resume\\":\\"function: 0x7e\\",\\"running\\":\\"function: 0x7f\\",\\"status\\":\\"function: 0x80\\",\\"wrap\\":\\"function: 0x81\\",\\"yield\\":\\"function: 0x82\\"}",
        "variablesReference": 1004,
      },
      Object {
        "name": "debug",
        "type": "object",
        "value": "{\\"debug\\":\\"function: 0x85\\",\\"gethook\\":\\"function: 0x87\\",\\"getinfo\\":\\"function: 0x88\\",\\"getlocal\\":\\"function: 0x89\\",\\"getmetatable\\":\\"function: 0x8b\\",\\"getregistry\\":\\"function: 0x8a\\",\\"getupvalue\\":\\"function: 0x8c\\",\\"getuservalue\\":\\"function: 0x86\\",\\"sethook\\":\\"function: 0x90\\",\\"setlocal\\":\\"function: 0x91\\",\\"setmetatable\\":\\"function: 0x92\\",\\"setupvalue\\":\\"function: 0x93\\",\\"setuservalue\\":\\"function: 0x8f\\",\\"traceback\\":\\"function: 0x94\\",\\"upvalueid\\":\\"function: 0x8e\\",\\"upvaluejoin\\":\\"function: 0x8d\\"}",
        "variablesReference": 1005,
      },
      Object {
        "name": "dofile",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "error",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "getmetatable",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "global_value",
        "type": "number",
        "value": "1",
        "variablesReference": 0,
      },
      Object {
        "name": "io",
        "type": "object",
        "value": "{\\"close\\":\\"function: 0xa3\\",\\"flush\\":\\"function: 0xa4\\",\\"input\\":\\"function: 0xa5\\",\\"lines\\":\\"function: 0xa6\\",\\"open\\":\\"function: 0xa7\\",\\"output\\":\\"function: 0xa8\\",\\"popen\\":\\"function: 0xa9\\",\\"read\\":\\"function: 0xaa\\",\\"stderr\\":\\"file (0x9220)\\",\\"stdin\\":\\"file (0x92b0)\\",\\"stdout\\":\\"file (0x9188)\\",\\"tmpfile\\":\\"function: 0xab\\",\\"type\\":\\"function: 0xac\\",\\"write\\":\\"function: 0xad\\"}",
        "variablesReference": 1006,
      },
      Object {
        "name": "ipairs",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "load",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "loadfile",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "math",
        "type": "object",
        "value": "{\\"abs\\":\\"function: 0xb6\\",\\"acos\\":\\"function: 0xb7\\",\\"asin\\":\\"function: 0xb8\\",\\"atan\\":\\"function: 0xb9\\",\\"ceil\\":\\"function: 0xba\\",\\"cos\\":\\"function: 0xbb\\",\\"deg\\":\\"function: 0xbc\\",\\"exp\\":\\"function: 0xbd\\",\\"floor\\":\\"function: 0xbf\\",\\"fmod\\":\\"function: 0xc0\\",\\"huge\\":\\"Infinity\\",\\"log\\":\\"function: 0xc2\\",\\"max\\":\\"function: 0xc3\\",\\"maxinteger\\":9223372036854776000,\\"min\\":\\"function: 0xc4\\",\\"mininteger\\":-9223372036854776000,\\"modf\\":\\"function: 0xc5\\",\\"pi\\":3.141592653589793,\\"rad\\":\\"function: 0xc6\\",\\"random\\":\\"function: 0xc7\\",\\"randomseed\\":\\"function: 0xc8\\",\\"sin\\":\\"function: 0xc9\\",\\"sqrt\\":\\"function: 0xca\\",\\"tan\\":\\"function: 0xcb\\",\\"tointeger\\":\\"function: 0xbe\\",\\"type\\":\\"function: 0xcc\\",\\"ult\\":\\"function: 0xc1\\"}",
        "variablesReference": 1007,
      },
      Object {
        "name": "next",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "os",
        "type": "object",
        "value": "{\\"clock\\":\\"function: 0xd5\\",\\"date\\":\\"function: 0xd6\\",\\"difftime\\":\\"function: 0xd7\\",\\"execute\\":\\"function: 0xd8\\",\\"exit\\":\\"function: 0xd9\\",\\"getenv\\":\\"function: 0xda\\",\\"remove\\":\\"function: 0xdb\\",\\"rename\\":\\"function: 0xdc\\",\\"setlocale\\":\\"function: 0xdd\\",\\"time\\":\\"function: 0xde\\",\\"tmpname\\":\\"function: 0xdf\\"}",
        "variablesReference": 1008,
      },
      Object {
        "name": "package",
        "type": "object",
        "value": "{\\"config\\":\\"/\\\\n;\\\\n?\\\\n!\\\\n-\\\\n\\",\\"cpath\\":\\"/usr/local/lib/lua/5.3/?.so;/usr/local/lib/lua/5.3/loadall.so;./?.so\\",\\"loaded\\":{\\"table\\":\\"0x0\\"},\\"loadlib\\":\\"function: 0xd2\\",\\"path\\":\\"/usr/local/share/lua/5.3/?.lua;/usr/local/share/lua/5.3/?/init.lua;/usr/local/lib/lua/5.3/?.lua;/usr/local/lib/lua/5.3/?/init.lua;./?.lua;./?/init.lua\\",\\"preload\\":{\\"table\\":\\"0x0\\"},\\"searchers\\":{\\"table\\":\\"0x0\\"},\\"searchpath\\":\\"function: 0xd3\\"}",
        "variablesReference": 1009,
      },
      Object {
        "name": "pairs",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "pcall",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "print",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "rawequal",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "rawget",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "rawlen",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "rawset",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "require",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "select",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "setmetatable",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "string",
        "type": "object",
        "value": "{\\"byte\\":\\"function: 0xe4\\",\\"char\\":\\"function: 0xe5\\",\\"dump\\":\\"function: 0xe6\\",\\"find\\":\\"function: 0xe7\\",\\"format\\":\\"function: 0xe8\\",\\"gmatch\\":\\"function: 0xe9\\",\\"gsub\\":\\"function: 0xea\\",\\"len\\":\\"function: 0xeb\\",\\"lower\\":\\"function: 0xec\\",\\"match\\":\\"function: 0xed\\",\\"pack\\":\\"function: 0xf2\\",\\"packsize\\":\\"function: 0xf3\\",\\"rep\\":\\"function: 0xee\\",\\"reverse\\":\\"function: 0xef\\",\\"sub\\":\\"function: 0xf0\\",\\"unpack\\":\\"function: 0xf4\\",\\"upper\\":\\"function: 0xf1\\"}",
        "variablesReference": 1010,
      },
      Object {
        "name": "table",
        "type": "object",
        "value": "{\\"concat\\":\\"function: 0xf5\\",\\"insert\\":\\"function: 0xf6\\",\\"move\\":\\"function: 0xfa\\",\\"pack\\":\\"function: 0xf7\\",\\"remove\\":\\"function: 0xf9\\",\\"sort\\":\\"function: 0xfb\\",\\"unpack\\":\\"function: 0xf8\\"}",
        "variablesReference": 1011,
      },
      Object {
        "name": "tonumber",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "tostring",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "type",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
      Object {
        "name": "utf8",
        "type": "object",
        "value": "{\\"char\\":\\"function: 0xff\\",\\"charpattern\\":\\"[\\",\\"codepoint\\":\\"function: 0xfe\\",\\"codes\\":\\"function: 0x101\\",\\"len\\":\\"function: 0x100\\",\\"offset\\":\\"function: 0xfd\\"}",
        "variablesReference": 1012,
      },
      Object {
        "name": "xpcall",
        "type": "string",
        "value": "function: 0x0",
        "variablesReference": 0,
      },
    ],
  },
  "command": "variables",
  "request_seq": 7,
  "seq": 10,
  "success": true,
  "type": "response",
}
`;

exports[`Lua Debug Adapter local check local_values 1`] = `
Object {
  "body": Object {
    "variables": Array [
      Object {
        "name": "local_value1",
        "type": "number",
        "value": "1",
        "variablesReference": 0,
      },
      Object {
        "name": "local_value2",
        "type": "string",
        "value": "\\"abc\\"",
        "variablesReference": 0,
      },
      Object {
        "name": "local_value3",
        "type": "number",
        "value": "1",
        "variablesReference": 0,
      },
      Object {
        "name": "local_value4",
        "type": "object",
        "value": "[4234.3]",
        "variablesReference": 1003,
      },
    ],
  },
  "command": "variables",
  "request_seq": 7,
  "seq": 10,
  "success": true,
  "type": "response",
}
`;

exports[`Lua Debug Adapter upvalues check upvalues 1`] = `
Object {
  "body": Object {
    "variables": Array [
      Object {
        "name": "a",
        "type": "number",
        "value": "1",
        "variablesReference": 0,
      },
      Object {
        "name": "t",
        "type": "object",
        "value": "[{\\"table\\":\\"0x0\\"},5,6]",
        "variablesReference": 1003,
      },
    ],
  },
  "command": "variables",
  "request_seq": 7,
  "seq": 10,
  "success": true,
  "type": "response",
}
`;
